File: bot/fetch_job_details.js
import puppeteer from 'puppeteer-extra';
import StealthPlugin from 'puppeteer-extra-plugin-stealth';
import fs from 'fs';
import path from 'path';
import os from 'os';

process.on('uncaughtException', (err) => {
    if (err.message && err.message.includes('Requesting main frame too early')) return;
    console.error(`Uncaught Exception: ${err.message}`);
});

process.on('unhandledRejection', (reason, promise) => {
    if (reason && reason.message && reason.message.includes('Requesting main frame too early')) return;
});

puppeteer.use(StealthPlugin());

(async () => {
    let browser;
    process.on('SIGTERM', async () => {
        if (browser) await browser.close();
        process.exit(0);
    });

    const args = process.argv.slice(2);
    if (args.length === 0) {
        console.error(JSON.stringify({ status: 'failed', message: 'No input provided' }));
        process.exit(1);
    }

    let inputData;
    try {
        inputData = JSON.parse(args[0]);
    } catch (e) {
        console.error(JSON.stringify({ status: 'failed', message: 'Invalid JSON input' }));
        process.exit(1);
    }

    const { url, platform, session_dir } = inputData;

    try {
        const launchOptions = { 
            headless: 'new', 
            defaultViewport: null,
            env: {
                ...process.env,
                XDG_CONFIG_HOME: os.tmpdir(),
                XDG_CACHE_HOME: os.tmpdir()
            }
        };
        
        if (process.env.DOCKER_ENV || inputData.is_docker) {
            launchOptions.args = [
                '--no-sandbox', '--disable-setuid-sandbox', '--disable-dev-shm-usage',
                '--disable-crash-reporter', '--disable-gpu', '--disable-software-rasterizer'
            ];
            launchOptions.userDataDir = '/tmp/puppeteer_data';
            launchOptions.executablePath = process.env.PUPPETEER_EXECUTABLE_PATH || '/usr/bin/chromium';
        }
        
        if (process.env.PUPPETEER_EXECUTABLE_PATH) {
            launchOptions.executablePath = process.env.PUPPETEER_EXECUTABLE_PATH;
        } else {
            const possibleChromePaths = [
                'C:\\Program Files\\Google\\Chrome\\Application\\chrome.exe',
                'C:\\Program Files (x86)\\Google\\Chrome\\Application\\chrome.exe',
                '/usr/bin/chromium', '/usr/bin/chromium-browser', '/usr/bin/google-chrome'
            ];
            for (const p of possibleChromePaths) {
                if (fs.existsSync(p)) { launchOptions.executablePath = p; break; }
            }
        }

        browser = await puppeteer.launch(launchOptions);
        const page = await browser.newPage();
        
        // Inject cookies if session exists
        if (session_dir) {
            const cookieFile = path.join(session_dir, 'cookies.json');
            if (fs.existsSync(cookieFile)) {
                try {
                    const cookies = JSON.parse(fs.readFileSync(cookieFile, 'utf8'));
                    await page.setCookie(...cookies);
                } catch (e) {}
            }
        }

        await page.setUserAgent('Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36');
        
        let targetUrl = url;
        if (url.includes('linkedin.com/jobs')) {
            const viewMatch = url.match(/view\/(\d+)/);
            const idMatch = url.match(/currentJobId=(\d+)/);
            const jobId = viewMatch ? viewMatch[1] : (idMatch ? idMatch[1] : null);
            if (jobId) {
                targetUrl = `https://www.linkedin.com/jobs-guest/jobs/api/jobPosting/${jobId}`;
            }
        }

        await page.goto(targetUrl, { waitUntil: 'networkidle2', timeout: 45000 });
        await new Promise(r => setTimeout(r, 3000)); // Extra wait for React/Angular hydration
        
        // Scroll to load dynamic content
        for (let i = 0; i < 3; i++) {
            await page.evaluate((step) => window.scrollTo(0, document.body.scrollHeight / 3 * (step+1)), i);
            await new Promise(r => setTimeout(r, 1000));
        }

        const details = await page.evaluate(async (plat) => {
            let description = '';
            let skills = '';
            
            // Helper function to format list items with bullets before getting innerText
            const formatNodeText = (node) => {
                if (!node) return '';
                const listItems = node.querySelectorAll('li');
                listItems.forEach(li => {
                    const text = li.innerText.trim();
                    if (text && !text.startsWith('•') && !text.startsWith('-')) {
                        li.innerText = '• ' + li.innerText;
                    }
                });
                return node.innerText.trim();
            };
            
            // Retry loop inside browser context for slow SPAs
            for (let attempt = 0; attempt < 5; attempt++) {
                if (plat === 'LINKEDIN' || window.location.href.includes('linkedin.com')) {
                    const descNode = document.querySelector('.jobs-description-content__text, .show-more-less-html__markup, .description__text');
                    if (descNode) description = formatNodeText(descNode);
                    if (!description) description = document.body.innerText.trim();
                } else if (plat === 'INDEED') {
                    const descNode = document.querySelector('#jobDescriptionText');
                    if (descNode) description = formatNodeText(descNode);
                } else {
                    // Try specific selectors for other known job boards
                    const specificSelectors = [
                        '.job-desc', 'section.job-desc', '.dang-inner-html', 
                        '[class*="styles_job-desc-container__"]', '[class*="styles_JDC__"]', '[class*="JobDescription"]', // Naukri
                        '.job-detail-content', // Uplers
                        '.competition-details', '.detail-container', // Unstop
                        '.job-details-wrapper', // Hirist
                        '.job-desc-content', '.job-details-container', // Cutshort
                        '#content', 'div[id="content"]', '.job__description', // Greenhouse
                        '.section-wrapper.page-full-width', // Lever
                        '#JobDescriptionContainer', '.jobDescriptionContent', // Glassdoor
                        '[data-automation-id="jobPostingDescription"]', // Workday
                        '#about-role', '[data-test="JobProfileAbout"]', // Wellfound / AngelList
                    ];

                    for (const selector of specificSelectors) {
                        const node = document.querySelector(selector);
                        if (node && node.innerText.length > 100) {
                            description = formatNodeText(node);
                            break;
                        }
                    }
                    
                    if (!description) {
                        // Also try matching by partial class name (for hashed modules like Naukri)
                        const partialNodes = document.querySelectorAll('[class*="job-desc"], [class*="JobDesc"]');
                        for (const pNode of partialNodes) {
                            if (pNode && pNode.innerText.length > 200) {
                                description = formatNodeText(pNode);
                                break;
                            }
                        }
                    }

                    // Fallback: Try to grab main content body
                    if (!description) {
                        const contentNodes = document.querySelectorAll('main, article, .job-description, .description, #job-details');
                        for (const node of contentNodes) {
                            if (node && node.innerText.length > 200) {
                                description = formatNodeText(node);
                                break;
                            }
                        }
                    }
                    
                    // Absolute fallback to body text if no structured container is found
                    if (!description) {
                        const bodyText = document.body.innerText.trim();
                        if (bodyText.length > 200) {
                            description = bodyText.substring(0, 5000); // Grab up to 5k chars to avoid crashing
                        }
                    }
                }
                
                if (description && description.length > 100) {
                    break;
                }
                
                // Wait 1 second before retrying inside browser context
                await new Promise(r => setTimeout(r, 1000));
            }

            return { description, skills };
        }, platform);
        
        await browser.close();
        
        console.log(JSON.stringify({ 
            status: 'success', 
            details: details 
        }));
        
    } catch (error) {
        if (browser) await browser.close();
        console.error(JSON.stringify({ status: 'failed', message: error.message }));
        process.exit(1);
    }
})();

File: bot/fetch_job_details.py
import sys
import json
import os
import time
import re
from playwright.sync_api import sync_playwright

def format_node_text(page, selector):
    return page.evaluate("""(sel) => {
        const node = document.querySelector(sel);
        if (!node) return '';
        const listItems = node.querySelectorAll('li');
        listItems.forEach(li => {
            const text = li.innerText.trim();
            if (text && !text.startsWith('•') && !text.startsWith('-')) {
                li.innerText = '• ' + li.innerText;
            }
        });
        return node.innerText.trim();
    }""", selector)

def format_node_text_by_element(page, element_handle):
    if not element_handle: return ''
    return element_handle.evaluate("""(node) => {
        if (!node) return '';
        const listItems = node.querySelectorAll('li');
        listItems.forEach(li => {
            const text = li.innerText.trim();
            if (text && !text.startsWith('•') && !text.startsWith('-')) {
                li.innerText = '• ' + li.innerText;
            }
        });
        return node.innerText.trim();
    }""")

def main():
    if len(sys.argv) < 2:
        print(json.dumps({'status': 'failed', 'message': 'No input provided'}))
        sys.exit(1)

    try:
        input_data = json.loads(sys.argv[1])
    except Exception as e:
        print(json.dumps({'status': 'failed', 'message': 'Invalid JSON input'}))
        sys.exit(1)

    url = input_data.get('url')
    platform = input_data.get('platform', '')
    session_dir = input_data.get('session_dir')

    if not url:
        print(json.dumps({'status': 'failed', 'message': 'Missing URL'}))
        sys.exit(1)

    target_url = url
    if 'linkedin.com/jobs' in url:
        view_match = re.search(r'view/(\d+)', url)
        id_match = re.search(r'currentJobId=(\d+)', url)
        job_id = view_match.group(1) if view_match else (id_match.group(1) if id_match else None)
        if job_id:
            target_url = f'https://www.linkedin.com/jobs-guest/jobs/api/jobPosting/{job_id}'

    try:
        with sync_playwright() as p:
            args = []
            if os.environ.get('DOCKER_ENV') or input_data.get('is_docker'):
                args = [
                    '--no-sandbox', '--disable-setuid-sandbox', '--disable-dev-shm-usage',
                    '--disable-crash-reporter', '--disable-gpu', '--disable-software-rasterizer'
                ]
            
            browser = p.chromium.launch(headless=True, args=args)
            context = browser.new_context(
                user_agent='Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36'
            )

            # stealth setup (if using stealth package, typically requires playwright-stealth)
            try:
                from playwright_stealth import stealth_sync
                stealth_sync(context)
            except ImportError:
                pass # Playwright stealth not available or failed to import

            if session_dir:
                cookie_file = os.path.join(session_dir, 'cookies.json')
                if os.path.exists(cookie_file):
                    try:
                        with open(cookie_file, 'r', encoding='utf-8') as f:
                            cookies = json.load(f)
                            # playwright accepts a list of dicts for cookies
                            context.add_cookies(cookies)
                    except Exception:
                        pass

            page = context.new_page()

            page.goto(target_url, wait_until='domcontentloaded', timeout=45000)
            page.wait_for_timeout(3000)

            for i in range(3):
                page.evaluate(f"window.scrollTo(0, document.body.scrollHeight / 3 * {i+1})")
                page.wait_for_timeout(1000)

            description = ''
            skills = ''

            for attempt in range(5):
                if platform == 'LINKEDIN' or 'linkedin.com' in page.url:
                    description = format_node_text(page, '.jobs-description-content__text, .show-more-less-html__markup, .description__text')
                    if not description:
                        description = page.evaluate("document.body.innerText.trim()")
                elif platform == 'INDEED':
                    description = format_node_text(page, '#jobDescriptionText')
                else:
                    specific_selectors = [
                        '.job-desc', 'section.job-desc', '.dang-inner-html', 
                        '[class*="styles_job-desc-container__"]', '[class*="styles_JDC__"]', '[class*="JobDescription"]', # Naukri
                        '.job-detail-content', # Uplers
                        '.competition-details', '.detail-container', # Unstop
                        '.job-details-wrapper', # Hirist
                        '.job-desc-content', '.job-details-container', # Cutshort
                        '#content', 'div[id="content"]', '.job__description', # Greenhouse
                        '.section-wrapper.page-full-width', # Lever
                        '#JobDescriptionContainer', '.jobDescriptionContent', # Glassdoor
                        '[data-automation-id="jobPostingDescription"]', # Workday
                        '#about-role', '[data-test="JobProfileAbout"]', # Wellfound / AngelList
                    ]

                    for selector in specific_selectors:
                        try:
                            node = page.locator(selector).first
                            if node.count() > 0 and len(node.inner_text() or '') > 100:
                                description = format_node_text_by_element(page, node.element_handle())
                                break
                        except Exception:
                            continue
                    
                    if not description:
                        try:
                            nodes = page.locator('[class*="job-desc"], [class*="JobDesc"]').all()
                            for node in nodes:
                                if len(node.inner_text() or '') > 200:
                                    description = format_node_text_by_element(page, node.element_handle())
                                    break
                        except Exception:
                            pass

                    if not description:
                        try:
                            nodes = page.locator('main, article, .job-description, .description, #job-details').all()
                            for node in nodes:
                                if len(node.inner_text() or '') > 200:
                                    description = format_node_text_by_element(page, node.element_handle())
                                    break
                        except Exception:
                            pass
                    
                    if not description:
                        try:
                            body_text = page.evaluate("document.body.innerText.trim()")
                            if len(body_text) > 200:
                                description = body_text[:5000]
                        except Exception:
                            pass

                if description and len(description) > 100:
                    break

                page.wait_for_timeout(1000)

            details = {
                'description': description,
                'skills': skills
            }

            browser.close()
            print(json.dumps({'status': 'success', 'details': details}))

    except Exception as e:
        print(json.dumps({'status': 'failed', 'message': str(e)}))
        sys.exit(1)

if __name__ == '__main__':
    main()
